pkg/api/bzzfile.go

```diff
@@ -0,0 +1,245 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package api
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"mime"
+	"mime/multipart"
+	"net/http"
+	"strconv"
+
+	"github.com/ethersphere/bee/pkg/collection/entry"
+	"github.com/ethersphere/bee/pkg/file"
+	"github.com/ethersphere/bee/pkg/file/joiner"
+	"github.com/ethersphere/bee/pkg/jsonhttp"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/gorilla/mux"
+)
+
+const (
+	MultiPartFormData = "multipart/form-data"
+)
+
+type FileUploadResponse struct {
+	Reference swarm.Address `json:"reference"`
+}
+
+// bzzFileUploadHandler uploads the file and its metadata supplied as a multipart http message.
+func (s *server) bzzFileUploadHandler(w http.ResponseWriter, r *http.Request) {
+	contentType, params, err := mime.ParseMediaType(r.Header.Get("Content-Type"))
+	if contentType != MultiPartFormData {
+		s.Logger.Debugf("bzz-file: no mutlipart: %v", err)
+		s.Logger.Error("bzz-file: no mutlipart")
+		jsonhttp.BadRequest(w, "not a mutlipart/form-data")
+		return
+	}
+
+	mr := multipart.NewReader(r.Body, params["boundary"])
+	for {
+		part, err := mr.NextPart()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			s.Logger.Debugf("bzz-file: read mutlipart: %v", err)
+			s.Logger.Error("bzz-file: read mutlipart")
+			jsonhttp.BadRequest(w, "error reading a mutlipart/form-data")
+			return
+		}
+
+		ctx := context.Background()
+
+		// find file size
+		fileSizeString := part.Header.Get("Content-Length")
+		if fileSizeString == "" {
+			s.Logger.Debugf("bzz-file: content length: %v", err)
+			s.Logger.Error("bzz-file: content length")
+			jsonhttp.BadRequest(w, "content length header missing")
+			return
+		}
+		fileSize, err := strconv.ParseUint(fileSizeString, 10, 64)
+		if err != nil {
+			s.Logger.Debugf("bzz-file: content length: %v", err)
+			s.Logger.Error("bzz-file: content length")
+			jsonhttp.BadRequest(w, "error parsing content length")
+			return
+		}
+
+		// first store the file and get its reference
+		fr, err := s.storePartData(ctx, part, fileSize)
+		if err != nil {
+			s.Logger.Debugf("bzz-file: file store: %v,", err)
+			s.Logger.Error("bzz-file: file store")
+			jsonhttp.InternalServerError(w, "could not store file data")
+			return
+		}
+
+		// try to find filename
+		// 1) in part header params
+		// 2) as formname
+		// 3) file reference hash
+		fileName := part.FileName()
+		if fileName == "" {
+			fileName = part.FormName()
+		}
+		if fileName == "" {
+			fileName = fr.String()
+		}
+		m := entry.NewMetadata(fileName)
+
+		// then find out content type
+		contentType := part.Header.Get("Content-Type")
+		if contentType == "" {
+			reader := bufio.NewReader(part)
```

Comment:
`part` is already read and here cannot be read any further as it reached EOF. The whole content type detection has to be done before any Read call on `part`. Basically, before storePartData call, the whole code block from line 85 to line 111 has to be moved, with one important difference. If the bufio.Reader is constructed it has to be passed to storePartData, not directly the `part` variable as `part` reader already moved its cursor, while the bufio reader did not, it just peeked, so it can read from the start, as `part` reader has advanced 512 bytes. The best way to do that is to create `var reader io.Reader = part` before `if contentType == "" {` line and here to do `reader = bufio.NewReader(part)` (assign to the variable, not create a new one), and use that `reader` variable in storePartData `fr, err := s.storePartData(ctx, reader, fileSize)`.

It looks to me that `contentType == ""` block is not tested and it would be good to have that test.